### DISCUSSION TITLE
Ideas for preventing memory leaks when hash used in `output.filename`

### DISCUSSION THREAD

**Discussion by srmagura** [webpack/webpack#16222] [Ideas]
My team was using:

```js
{
  /* ... */
  output: {
    filename: '[name].[chunkhash].js',
    clean: true
  },
  devtool: 'cheap-module-source-map'
}
```

This caused Webpack's memory usage to increase by ~300 MB on each rebuild. I eventually found [this comment](https://github.com/webpack/webpack/issues/12947#issuecomment-804007892) by @sokra:

> Best avoid using contenthashes during development. This does have similar problems in other places, e. g. it writes a lot files into the filesystem (on disk or in memory).

I removed the `output.filename` option, so that the default of `[name].js` is used, and this fixed the rapidly increasing memory usage.

My questions:

1. Since including `[fullhash]` / `[chunkhash]` / `[contenthash]` in `output.filename` is known to cause memory issues, could Webpack print a warning if `output.filename` contains one of these variables and `mode === 'development'`? If yes, then I can attempt a PR.
2. Could Webpack be changed so that it clears cached assets from old builds automatically? I don't know enough about Webpack internals to know if this is feasible.

Ref https://github.com/webpack/webpack/issues/12947

**Comment by alexander-akait** [comment]
How many modules do you have? Search and Replace hashes can consume more memories (but not ~300MB), do you have any non official plugins (will be great to see at configuration too)?

**Reply by srmagura** [reply]
I simplified the config until it looked like this:

```js
const path = require('path');

module.exports = {
  mode: 'development',
  entry: {
    app: path.resolve(__dirname, 'src/index.tsx'),
  },
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.json'],
    modules: [path.resolve(__dirname, '.'), 'node_modules'],
  },
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: require.resolve('swc-loader'),
      },
    ],
  },
  output: {
    filename: '[name].[chunkhash].js', // removing this line fixed it

    path: path.resolve(__dirname, 'dist'),
    publicPath: '/',
    clean: true,
    devtoolModuleFilenameTemplate: 'webpack:///[absolute-resource-path]',
  },
  devtool: 'cheap-module-source-map', // removing this line also seemed to fix it? Not sure if there is a connection there
};
```

The only third party plugin is swc. I tried replacing it with Babel, and the memory leak persisted, which led me to conclude that swc was not the problem.

I am not sure how to determine the exact number of modules, but I imagine the number is quite high. I am attaching the results of running a single development build with `stats: 'verbose'` — hopefully that is helpful.

If you think this is a bug, I can try to prepare a reproduction (though the problem is occurring on a closed source codebase so I would need to strip out our real code while still demonstrating the memory issue.)

**Reply by alexander-akait** [reply]
Can you try to set to `false` https://webpack.js.org/configuration/optimization/#optimizationrealcontenthash and check memory usage?

**Reply by srmagura** [reply]
I set `realContentHash: false` and the memory usage still grew rapidly.

Edit: also realized I forgot to attach those logs. Will do it when I get back to my home PC.

**Reply by alexander-akait** [reply]
Feel free to ping me

**Reply by srmagura** [reply]
Hey @alexander-akait, here are the verbose logs! This was from running our full webpack config (not the simplified config I showed earlier). Hopefully this gives you an idea of the total number of modules we have. Thank you 😊

[spot-webpack-verbose.zip](https://github.com/webpack/webpack/files/9526416/spot-webpack-verbose.zip)

**Reply by alexander-akait** [reply]
Oh, you have really big application :smile: Can you check that what using `[fullhash]`/`[contenthash]` (any value except `chunkhash`) is leeking memory too? 

I found you use `worker-loader` too, there is some magic there. Why don't use built-in solution (i.e. `new Worker(new URL(...))`)? `worker-loader` is deprecated and uses the old approach (which potential has memory leading in some rare cases). 

Ideally I recommend to disable loaders and plugins step by step to check that no other things are leaking (I know it requires time, but we should check it out first). Also can you disable source maps and check too?

**Reply by srmagura** [reply]
@alexander-akait 

> Ideally I recommend to disable loaders and plugins step by step to check that no other things are leaking (I know it requires time, but we should check it out first).

I did do this (posted the minimal config earlier), but the logs I sent you are from running Webpack with our full config.

Just wondering, are you asking me to try these different things because you think there could be a bug in Webpack? If so, I will try them out and report back to you. If you are just trying to help solve my specific issue, there is no need since the issue was resolved once I removed `chunkhash`.

Thanks for mentioning the `worker-loader` thing. I found out it is being brought in by `pdfjs-dist/webpack`, rather than our Webpack config.

**Reply by alexander-akait** [reply]
> Just wondering, are you asking me to try these different things because you think there could be a bug in Webpack? If so, I will try them out and report back to you. If you are just trying to help solve my specific issue, there is no need since the issue was resolved once I removed chunkhash.

The problem is that there are a lot of places where we can have a bug - non official loaders/plugins and including in the webpack itself. The strange circumstance is that the problem is only `[chunkhash]`. Feeling like some kind of regex try to parser filename, usually it makes a spike in memory consumption (but that's just my guess). Maybe you can profile and capture memory usage (here example of the another memory problem - https://github.com/webpack/webpack/issues/12771)?
